src/accumulate.jl

```diff
@@ -2,104 +2,160 @@
+        end
+        if lane > 4
+            val = op(cache[thread - 4], val)
+            sync_warp(mask >> 4)
+            cache[thread] = val
+            sync_warp(mask >> 4)
+        end
+        if lane > 8
+            val = op(cache[thread - 8], val)
+            sync_warp(mask >> 8)
+            cache[thread] = val
+            sync_warp(mask >> 8)
+        end
+        if lane > 16
+            val = op(cache[thread - 16], val)
+        end
+    end
+    return val
+end
+
+function partial_scan!(op::Function, output::CuDeviceArray{T}, input::CuDeviceArray,
+                      aggregates::Union{Nothing, CuDeviceArray{T}}, Rdim, Rpre,
+                      Rpost, Rother, init, ::Val{inclusive}=Val(true),
+                      ::Val{shuffle}=Val(true)) where {T, inclusive, shuffle}
     threads = blockDim().x
     thread = threadIdx().x
     block = blockIdx().x
 
-    temp = @cuDynamicSharedMem(T, (2*threads,))
+    # wid: warp index in block, lane: lane index in warp
+    wid, lane = fldmod1(thread, warpsize())
+    exclusive = !inclusive
+
+    # cache: storage for non shuffle kernels
+    partials = @cuDynamicSharedMem(T, 32)
+    cache = @cuDynamicSharedMem(T, shuffle ? 0 : threads, offset=sizeof(partials))
 
     # iterate the main dimension using threads and the first block dimension
     i = (blockIdx().x-1) * blockDim().x + threadIdx().x
     # iterate the other dimensions using the remaining block dimensions
     j = (blockIdx().z-1) * gridDim().y + blockIdx().y
 
-    if j > length(Rother)
-        return
-    end
+    j > length(Rother) && return
 
     @inbounds begin
         I = Rother[j]
         Ipre = Rpre[I[1]]
         Ipost = Rpost[I[2]]
-    end
 
-    # load input into shared memory (apply `op` to have the correct type)
-    @inbounds temp[thread] = if i <= length(Rdim)
-        op(neutral, input[Ipre, i, Ipost])
-    else
-        op(neutral, neutral)
-    end
+        # convert input val to correct type
+        # take element from previous index in exclusive scan
+        exclusive && (i = i - 1)
+        value = if i <= length(Rdim) && i != 0
+            convert(T, input[Ipre, i, Ipost])
+        else
+            convert(T, input[Ipre, 1, Ipost])
+        end
+        exclusive && (i = i + 1)
 
-    # build sum in place up the tree
-    offset = 1
-    d = threads>>1
-    while d > 0
-        sync_threads()
-        @inbounds if thread <= d
-            ai = offset * (2*thread-1)
-            bi = offset * (2*thread)
-            temp[bi] = op(temp[ai], temp[bi])
+        # apply init to the first element
+        if i == 1 && init !== nothing
+            value = inclusive ? op(init, value) : init
         end
-        offset *= 2
-        d >>= 1
-    end
 
-    # clear the last element
-    @inbounds if thread == 1
-        temp[threads] = neutral
-    end
+        value = if shuffle
+            scan_warp(op, value, lane)
+        else
+            cache[thread] = value
```

Comment:
Try adding a `sync_warp()` here. I didn't get any errors wrt sync_check/race_check so I purposely left it out even though it's wrong.